model_analyzer/triton/server.py

```diff
@@ -0,0 +1,309 @@
+    A config class to set arguments to the Triton Inference
+    Server. An argument set to None will use the server default.
+    """
+    
+    def __init__(self):
+        # Args will be a dict with the string representation as key
+        self._server_args = {
+            # Logging
+            'log-verbose' : None,
+            'log-info' : None,
+            'log-warning' : None,
+            'log-error' : None,
+            'id' : None,
+        
+            # Model Repository
+            'model-store' : None,
+            'model-repository' : None,
+            
+            # Exit
+            'exit-timeout-secs' : None,
+            'exit-on-error' : None,
+
+            # Strictness
+            'strict-model-config' : None,
+            'strict-readiness' : None,
+            
+            # API Servers
+            'allow-http' : None,
+            'http-port' : None,
+            'http-thread-count' : None,
+            'allow-grpc' : None,
+            'grpc-port' : None,
+            'grpc-infer-allocation-pool-size' : None,
+            'grpc-use-ssl' : None,
+            'grpc-server-cert' : None,
+            'grpc-server-key' : None,
+            'grpc-root-cert' : None,
+            'allow-metrics' : None,
+            'allow-gpu-metrics' : None,
+            'metrics-port' : None,
+            
+            # Tracing
+            'trace-file' : None,
+            'trace-level' : None,
+            'trace-rate' : None,
+            
+            # Model control
+            'model-control-mode' : None,
+            'repository-poll-secs' : None,
+            'load-model' : None,
+
+            # Memory and GPU
+            'pinned-memory-pool-byte-size' : None,
+            'cuda-memory-pool-byte-size' : None,
+            'min-supported-compute-capability' : None,
+            
+            # Backend config
+            'backend-directory' : None,
+            'backend-config' : None,
+            
+            'allow-soft-placement' : None,
+            'gpu-memory-fraction' : None,
+            'tensorflow-version' : None
+        }
+
+    def to_cli_string(self):
+        """
+        Utility function to convert a config into a
+        string of arguments to the server with CLI.
+        """
+        return ' '.join(['--{}={}'.format(key,val) if val else '' \
+                        for key,val in self._server_args.items()])
+
+    def __getitem__(self, key):
+        return self._server_args[key]
+    
+    def __setitem__(self, key, value):
+        if key in self._server_args:
+            self._server_args[key] = value
+        else:
+            raise Exception("The argument '{}' to the Triton Inference Server"
+                             "is not supported by the model analyzer.".format(key))
+    
+class TritonServerFactory:
+    """
+    A factory for creating TritonServer instances
+    """
+
+    def __init__(self):
+        pass
+
+    def create_server(self, run_type, model_path, version, config):
```

Comment:
This is a slight misuse of the Factory pattern here. It forces you to have the same signature for the different object. The correct way is to have separate create_x_server methods.  Here is a good example.

https://refactoring.guru/design-patterns/factory-method/python/example

The point of the factory method to separate the dependency from user of the interface and the classes that derive the interface.

Reply:
So the correct way according to that example method would be to have a `TritonHTTPClientFactory` and a `TritonGRPCClientFactory` derived from `TritonClientFactory`? Would it be enough to just have 2 static create methods in the `TritonClientFactory` class : `create_http_client` and `create_grpc_client`?

Reply:
It C++, that is the way. However, I think python does things a bit differently. For now, create the two separate methods. Note you will need to do the same thing in your other Factory class.  I will look more in to the Factory pattern in python to see if further refactoring is needed.

Reply:
In my latest commit I created separate classes, the advantage being that it is extensible without needing to change the base interface (as opposed to adding static create_x methods.